src/Microsoft.Azure.SignalR.Common/Utilities/AckHandler.cs

```diff
@@ -0,0 +1,99 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.SignalR
+{
+    internal sealed class AckHandler : IDisposable
+    {
+        private readonly ConcurrentDictionary<int, AckInfo> _acks = new ConcurrentDictionary<int, AckInfo>();
+        private readonly Timer _timer;
+        private readonly TimeSpan _ackInterval = TimeSpan.FromSeconds(5);
+        private volatile int _currentId = 0;
+
+        public AckHandler()
+        {
+            bool restoreFlow = false;
+            try
+            {
+                if (!ExecutionContext.IsFlowSuppressed())
+                {
+                    ExecutionContext.SuppressFlow();
+                    restoreFlow = true;
+                }
+
+                _timer = new Timer(state => ((AckHandler)state).CheckAcks(), state: this, dueTime: _ackInterval, period: _ackInterval);
+            }
+            finally
+            {
+                // Restore the current ExecutionContext
+                if (restoreFlow)
+                {
+                    ExecutionContext.RestoreFlow();
+                }
+            }
+        }
+
+        public Task<bool> CreateAck(out int id)
+        {
+            id = Interlocked.Increment(ref _currentId);
+            return _acks.GetOrAdd(id, _ => new AckInfo()).Tcs.Task;
+        }
+
+        public void TriggerAck(int id, bool isSuccess)
+        {
+            if (_acks.TryRemove(id, out var ack))
+            {
+                ack.Tcs.TrySetResult(isSuccess);
+            }
+        }
+
+        private void CheckAcks()
+        {
+            var utcNow = DateTime.UtcNow;
+
+            foreach (var pair in _acks)
+            {
+                if (utcNow > pair.Value.Expired)
+                {
+                    if (_acks.TryRemove(pair.Key, out var ack))
+                    {
+                        // If acks not coming back in time, do not throw an exception
+                        ack.Tcs.TrySetResult(false);
+                    }
+                }
+            }
+        }
+
+        public void Dispose()
```

Comment:
>Dispose [](start = 20, length = 7)

who to dispose it?

Reply:
Made containerbase disposable to dispose AckHandler and ready for scaling.